backend_jobs/ingestion_pipeline/pipeline_lib/database_functions.py

```diff
@@ -0,0 +1,143 @@
+    """
+    if not firebase_admin._apps:
+        firebase_admin.initialize_app(credentials.ApplicationDefault(), {
+        'projectId': 'step-project-ellispis',
+        })
+    return firestore.client()
+
+class UploadToDatabase(beam.DoFn):
+    """
+    Uploads asynchronously the information that was extracted from each
+    image to the project's database.
+    """
+    def setup(self):
+        self.database_firebase = initialize_database()
+
+    def process(self, element, provider, job_name):
+        """
+        Adds/Updates the project's database to contain documents with image attributes.
+        In addition it adds for each image a sub collection with information
+        on the pipelinerun and the provider.
+        Args:
+            element: ImageAttribute field
+            provider: the provider we are running now
+            job_name: the job name that we are running now
+        """
+        doc_ref = self.database_firebase.collection(IMAGES_COLLECTION).document(element.id)
+        doc = doc_ref.get()
+        sub_collection_ref= doc_ref.collection(IMAGES_SUB_COLLECTION)
+        sub_collection_doc_ref= sub_collection_ref.document()
+        if doc.exists:
+            #doc found- image has been ingested already
+            update_document(provider,doc, doc_ref,job_name)
+        else:
+            #doc not found- image has not been ingested already
+            add_document(element, provider, doc_ref)
+        #Adding a doc to the sub collection (pipelinerun) in the image collection
+        upload_sub_collection(element, provider, job_name, sub_collection_doc_ref)
+
+def add_document(element, provider, doc_ref):
+    geo_point_coordinates=cloud_firestore.GeoPoint(
+        float(element.coordinates['latitude']),
+        float(element.coordinates['longitude']))
+    doc_ref.set({
+        u'url': element.url,
+        u'ingestedProviders':[provider.provider_id],
+        u'ingestedRuns':[1],
+        u'coordinates': geo_point_coordinates,
+        u'dateIngested': datetime.now(),
+        u'dateShot': element.date_shot,
+        u'dateFields':get_date_fields(element.date_shot),
+        u'geoHashes': get_geo_hashes_map(element.coordinates),
+        u'imageAttributes':{
+            u'format': element.format,
+            u'resolution':element.resolution},
+        u'attribution': element.attribution,
+        u'random': random.randint(1,101),
+        u'visibility': provider.visibility.value,
+    })
+
+def update_document(provider,doc, doc_ref,job_name):
+    ingested_runs = doc.to_dict()[u'ingestedRuns']
+    ingested_runs.append(job_name)
+    ingested_providers = doc.to_dict()[u'ingestedProviders']
+    if provider.provider_id not in ingested_providers:
+        ingested_providers.append(provider.provider_id)
+    doc_ref.update({
+        u'ingestedRuns':ingested_runs,
+        u'ingestedProviders': ingested_providers,
+        u'visibility': max(provider.visibility, doc.to_dict()[u'visibility']),
```

Comment:
visibilit is an enum, we should not technically depend on its value directly (implementation detail). maybe consider a helper function on enum instead

Reply:
added helper function _get_max_visibility